src/lcs.rs

```diff
@@ -1,96 +1,79 @@
+use std::cmp::max;
+
+// strsplit is like `s.split(split)`, except that if `split` is "", it
+// trims the leading and trailing empty elements, since the `lcs`
+// logic won't handle those properly.
+fn strsplit<'a>(s: &'a str, split: &str) -> Vec<&'a str> {
+    let mut si = s.split(split);
+    if split == "" {
+        si.next();
+    }
+    let mut v: Vec<&str> = si.collect();
+    if split == "" {
+        v.pop();
+    }
+    v
+}
+
 // finds the longest common subsequences
 // outputs the edit distance and a string containing
 // all chars both inputs have in common
 #[allow(non_snake_case)]
 #[cfg_attr(feature = "cargo-clippy", allow(many_single_char_names))]
 pub fn lcs(orig: &str, edit: &str, split: &str) -> (i32, String) {
-
     // make list by custom splits
-    let a: Vec<&str> = orig.split(split).collect();
-    let b: Vec<&str> = edit.split(split).collect();
-
-    let N = a.len() as i32;
-    let M = b.len() as i32;
-
-    let MAX = N + M;
-
-    let mut v: Vec<i32> = (-MAX..MAX).collect();
+    let a = strsplit(orig, split);
+    let b = strsplit(edit, split);
 
-    // container to hold common subsequence
-    let mut common = String::new();
-
-    v[1] = 0;
-
-    // iterate over D = "edit steps"
-    for D in 0..MAX {
-        let mut max = 0;
-        let mut max_snake: Box<String> = Box::new("".to_string());
-
-        // TODO replace with
-        // for k in (-D..D+1).step_by(2) {
-        // once it's stable
-
-        let mut k = -D;
-
-        while k < D + 1 {
-            let mut snake = String::new();
-
-            let mut x;
-
-            let index = (MAX + k - 1) as usize;
-            if k == -D || k != D && v[index - 1] < v[index + 1] {
-                x = v[index + 1];
-            } else {
-                x = v[index - 1] + 1;
-            }
+    let N = a.len();
+    let M = b.len();
 
-            let mut y = x - k;
+    let mut idx: Vec<usize> = Vec::with_capacity(N * M);
+    idx.resize(N * M, 0);
 
-            while x < N && y < M && a[x as usize] == b[y as usize] {
-                if !snake.is_empty() {
-                    // add back the splits that were taken away
-                    snake.push_str(split);
+    for i in 0..N {
+        for j in 0..M {
+            if b[j] == a[i] {
+                if i == 0 || j == 0 {
```

Comment:
There's a bunch of explicit edge-case handling here in order to avoid needing an `(N+1)*(M+1)`-sized array; If you're willing to allocate a buffer row/column of zero values, a bunch of these can go away. Since we're allocating regardless, it might be worth eating the slight memory increase for a substantial code simplification.